balloon_stress: fix "background stress is not alive"

huangyum opened the pull request:
Signed-off-by: Yumei Huang <yuhuang@redhat.com>

huangyum commented:
Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_stress: PASS (4934.45 s)

Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.balloon_stress: PASS (2910.01 s)

huangyum commented:
I've uploaded the whole log to bugzilla. Following are the traceback part. Running more stress in guest make it hang, causing timeout error. 

 07:18:05 DEBUG| [stderr] Exception in thread Thread-7:
07:18:05 DEBUG| [stderr] Traceback (most recent call last):
07:18:05 DEBUG| [stderr]   File "/usr/lib64/python2.7/threading.py", line 812, in __bootstrap_inner
07:18:05 DEBUG| [stderr]     self.run()
07:18:05 DEBUG| [stderr]   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-62.0-py2.7.egg/virttest/utils_misc.py", line 108, in run
07:18:05 DEBUG| [stderr]     self._retval = self._target(*self._args, **self._kwargs)
07:18:05 DEBUG| [stderr]   File "/var/lib/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/balloon_stress.py", line 35, in run_stress
07:18:05 DEBUG| [stderr]     stress_bg.load_stress_tool()
07:18:05 DEBUG| [stderr]   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-62.0-py2.7.egg/virttest/utils_test/__init__.py", line 1835, in manage_session
07:18:05 DEBUG| [stderr]     return func(self)
07:18:05 DEBUG| [stderr]   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-62.0-py2.7.egg/virttest/utils_test/__init__.py", line 1901, in load_stress_tool
07:18:05 DEBUG| [stderr]     self.install()
07:18:05 DEBUG| [stderr]   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-62.0-py2.7.egg/virttest/utils_test/__init__.py", line 1999, in install
07:18:05 DEBUG| [stderr]     status, output = self.cmd_status_output(self.make_cmds)
07:18:05 DEBUG| [stderr]   File "/usr/lib/python2.7/site-packages/aexpect/client.py", line 1165, in cmd_status_output
07:18:05 DEBUG| [stderr]     o = self.cmd_output(cmd, timeout, internal_timeout, print_func, safe)
07:18:05 DEBUG| [stderr]   File "/usr/lib/python2.7/site-packages/aexpect/client.py", line 1086, in cmd_output
07:18:05 DEBUG| [stderr]     raise ShellTimeoutError(cmd, o)
07:18:05 DEBUG| [stderr] ShellTimeoutError: Timeout expired while waiting for shell command to complete: 'cd /home/stress-1.0.4/;./configure && make install'    (output: u'checking for a BSD-compatible install...

......

07:18:21 ERROR| Reproduced traceback from: /usr/lib/python2.7/site-packages/avocado_plugins_vt-62.0-py2.7.egg/avocado_vt/test.py:434
07:18:21 ERROR| Traceback (most recent call last):
07:18:21 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-62.0-py2.7.egg/virttest/error_context.py", line 135, in new_fn
07:18:21 ERROR|     return fn(*args, **kwargs)
07:18:21 ERROR|   File "/var/lib/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/balloon_stress.py", line 66, in run
07:18:21 ERROR|     test.error("Background stress process is not alive")
07:18:21 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_framework-61.0-py2.7.egg/avocado/core/test.py", line 1064, in error
07:18:21 ERROR|     raise exceptions.TestError(message)
07:18:21 ERROR| TestError: Background stress process is not alive

luckyh commented:
@huangyum thanks for the information.

So, from the error log we can see there are two issues:

1. the stress tool had not been installed properly (due to command timeout)

The problem is that 60s is not long enough to finish the installation in this case (especially by doing the installation in a thread).
```python
# avocado-vt/virttest/utils_test/__init__.py
    def install(self):
        """
        To download, abstract, build and install the stress tool
        """
        self.download_stress()
        install_path = os.path.join(self.dst_path, self.base_name, self.work_path)
        self.make_cmds = "cd %s;%s" % (install_path, self.make_cmds)
        logging.info('make and install the %s', self.stress_type)
        status, output = self.cmd_status_output(self.make_cmds)    # <-- timeout = 60s
        if status != 0:
            raise exceptions.TestError(
                "Installation failed with output:\n %s" % output)
```

2. the test run without waiting for stress process being launched

The property way to solve this problem is to check the process in a duration, if the process appears in the duration, go on the test, if not, reports error.

Based on the above facts, I don't think this patch could really fix the issues, so NACK.

huangyum commented:
@luckyh Thanks for your suggestion.  If you look through the whole log, you would find it is not the first time to install stress when error happens.  The former code repeats `stress_bg.load_stress_tool()` in an infinite loop, which make guest keep installing and running stress.  

In my test, when the error happens,  four stress processes are running, which consumes about 2G memory. And at that time, the balloon is about 2431M.  Apparently, guest hangs, even call trace.  So the fifth installation failed.

07:14:20 INFO | Launch stress with command: nohup stress --cpu 4 --io 4 --vm 2 --vm-bytes 256M &
07:14:20 DEBUG| Sending command (safe): nohup stress --cpu 4 --io 4 --vm 2 --vm-bytes 256M &

07:14:43 INFO | Launch stress with command: nohup stress --cpu 4 --io 4 --vm 2 --vm-bytes 256M &
07:14:43 DEBUG| Sending command (safe): nohup stress --cpu 4 --io 4 --vm 2 --vm-bytes 256M &

07:15:13 INFO | Launch stress with command: nohup stress --cpu 4 --io 4 --vm 2 --vm-bytes 256M &
07:15:13 DEBUG| Sending command (safe): nohup stress --cpu 4 --io 4 --vm 2 --vm-bytes 256M &

07:16:30 INFO | Launch stress with command: nohup stress --cpu 4 --io 4 --vm 2 --vm-bytes 256M &
07:16:30 DEBUG| Sending command (safe): nohup stress --cpu 4 --io 4 --vm 2 --vm-bytes 256M &

07:18:05 DEBUG| (monitor avocado-vt-vm1.qmpmonitor1) Response to 'query-balloon' (re-formated)
07:18:05 DEBUG| (monitor avocado-vt-vm1.qmpmonitor1)    actual: 2549088256
07:18:05 DEBUG| [stderr] Exception in thread Thread-7:
07:18:05 DEBUG| [stderr] Traceback (most recent call last):
......

I agree with your solution, and it is what I did.  I move the stress part out of the infinite loop, run it only once, and keep checking the stress process during the test.  It works for me and the test passes. Thanks!

huangyum commented:
@luckyh The tool is installed before balloon operation,  so it wouldn't be affected by min_sz. 

As for the second issue, I've updated the code to check the stress status in the loop.  For windows guest, I introduced mplayer as the video player and play the video in a continuous loop.

yanan-fu commented:
The stress method for linux and windows guest is ok for me.
My concern is why not use new thread to run stress test as before ?

yanan-fu commented on qemu/tests/balloon_stress.py:
```diff
@@ -51,8 +56,15 @@ def run_stress(test, params, env, vm):
+        video_url = params["video_url"]
+        play_video_cmd = params["play_video_cmd"] % (video_player, video_url)
+        session.cmd(play_video_cmd)
+        time.sleep(5)
```
How about wait_for check_bg_running ?

huangyum replied:
I'm not sure if it is necessary. wait_for means we will keep checking the process until it is running. But IMHO, we don't want to wait every time we send a cmd to guest. I would prefer to check the process only once before doing next test instead of waiting for it. Thanks!

yanan-fu replied:
So, what's the sleep here used for ?   It is used to wait for the stress is running after loaded (only one time) then begin to do balloon, right ?  I think wait for is more reliable than sleep.

Even use `wait_for`, we only check once before doing next step, not every time.

The latest update use `sleep` for both linux and windows.  For linux guest, `load_stress_tool` already wait for the stress running.  I think it is redundant for linux guest.

huangyum replied:
@yanan-fu If we only check once, why do we need wait_for ? It might be redundant for linux guest to do extra sleep , but I think it makes code more readable  to put it before check_bg_running instead of only putting it in the block for windows. Thanks!

yanan-fu replied:
@huangyum  your `once` is not what i mean. I said `once`, it means we only use `wait_for` one time to wait for the stress begin running.  I don't care how many times it checking the process in func `wait_for` until time out.  This is what `wait_for` used for.

huangyum replied:
@yanan-fu Thanks, have updated to use wait_for, please help review.  Thanks!

huangyum commented:
@yanan-fu The case intends to run a stress application in guest and keep it running during balloon test.  Now the stress application runs as a background job, we don't need to start a new thread to do that.

huangyum commented:
Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_stress: PASS
 Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_in_use.with_stop_continue.before_bg_test: PASS
Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.win_video_play: PASS

After apply internal patch #28, driver_load_stress.with_balloon.before_bg_test would pass as well.

huangyum commented:
@yanan-fu @lijinlijin Please help review again, I made more changes. Thanks!

yanan-fu commented on qemu/tests/cfg/balloon_in_use.cfg:
```diff
@@ -19,17 +19,10 @@
         i386:
             program_files = "%ProgramFiles%"
```
"program_files" is useless now, can be removed.

yanan-fu commented on qemu/tests/cfg/balloon_in_use.cfg:
```diff
@@ -14,22 +14,11 @@
-        kmplayer_install_cmd = "start /wait WIN_UTILS:\kmplayer\%s\KMPlayer-setup.exe /SP- /VERYSILENT"
-        kmplayer_path = "${program_files}\KMPlayer\kmplayer.exe"
+        mplayer_path = "WIN_UTILS:\mplayer.exe"
+        play_video_cmd = "start /MIN %s %s -loop 0 -fs"
```
Need a param "target_process" for cases "during_bg_test". Otherwise,it will always get True status when check_bg_running in driver_in_use.py.
```
 24     def check_bg_running(target_process):                                       
 32         session = vm.wait_for_login()                                           
 33         list_cmd = params.get("list_cmd", "wmic process get name")              
 34         output = session.cmd_output_safe(list_cmd, timeout=60)                  
 35         process = re.findall(target_process, output, re.M | re.I)               
 36         session.close()                                                         
 37         return bool(process)  
```

huangyum replied:
@yanan-fu Yeah, you are right, thanks!

huangyum commented:
@yanan-fu  @luckyh  pls help review again, thanks!
Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_stress: PASS (459.87 s)

yanan-fu commented:
Code LGTM and test pass:
```
python3 ConfigTest.py --testcase=balloon_in_use,balloon_stress,driver_load_stress.with_balloon,win_video_play --guestname=Win2016
 (01/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: PASS (650.76 s)
 (02/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_in_use.with_stop_continue.before_bg_test: PASS (1125.15 s)
 (03/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_in_use.with_stop_continue.during_bg_test: PASS (4605.10 s)
 (04/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_in_use.with_stop_continue.after_bg_test: PASS (583.13 s)
 (05/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_in_use.with_shutdown.before_bg_test: PASS (630.78 s)
 (06/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_in_use.with_shutdown.during_bg_test: PASS (766.95 s)
 (07/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_in_use.with_shutdown.after_bg_test: PASS (690.47 s)
 (08/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_in_use.with_reboot.before_bg_test: PASS (612.53 s)
 (09/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_in_use.with_reboot.during_bg_test: PASS (6114.61 s)
 (10/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_in_use.with_reboot.after_bg_test: PASS (670.26 s)
 (11/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_in_use.with_system_reset.before_bg_test: PASS (651.49 s)
 (12/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_in_use.with_system_reset.during_bg_test: PASS (5684.96 s)
 (13/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_in_use.with_system_reset.after_bg_test: PASS (658.69 s)
 (14/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_in_use.with_live_migration.before_bg_test: PASS (582.86 s)
 (15/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_in_use.with_live_migration.during_bg_test: PASS (4445.96 s)
 (16/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_in_use.with_live_migration.after_bg_test: PASS (613.42 s)
 (17/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.driver_load_stress.with_balloon.before_bg_test: PASS (546.85 s)
 (18/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.driver_load_stress.with_balloon.during_bg_test: PASS (1094.57 s)
 (19/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.driver_load_stress.with_balloon.after_bg_test: PASS (609.16 s)
 (20/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.win_video_play: PASS (284.96 s)
 (21/21) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.balloon_stress: PASS (15815.81 s)


Acked-by: Yanan Fu <yfu@redhat.com>

lijinlijin commented:
Is the mplayer.exe for 64-bit guests only?
Jobs passed on win7-64 guest, but all failed on win7-32 guest.
Following is the error msg:
10:11:25 ERROR| aexpect.exceptions.ShellCmdError: Shell command failed: 'start /MIN D:\\mplayer.exe http://fileshare.englab.nay.redhat.com/pub/section2/kvmauto/video/big_buck_bunny_480p_stereo.avi -loop 0 -fs'    (status: 216,    output: "This version of D:\\mplayer.exe is not compatible with the version of Windows you're running. Check your computer's system information to see whether you need a x86 (32-bit) or x64 (64-bit) version of the program, and then contact the software publisher.\n\n")

huangyum commented:
@lijinlijin I just updated the mplayer application in winutils.iso to be compatible with 32 bit windows guest. Test passed with win2008sp2 32 bit guest. Please help review again, thanks!

huangyum commented:
(1/5) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win2008.i386.sp2.io-github-autotest-qemu.win_video_play: PASS (425.68 s)
 (2/5) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win2008.x86_64.sp2.io-github-autotest-qemu.win_video_play: PASS (444.01 s)
 (3/5) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win2008.x86_64.r2.io-github-autotest-qemu.win_video_play: PASS (467.71 s)
 (4/5) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win7.i386.sp1.io-github-autotest-qemu.win_video_play: PASS (488.74 s)
 (5/5) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win7.x86_64.sp1.io-github-autotest-qemu.win_video_play: PASS (460.50 s)

lijinlijin commented:
Acked-by: Li Jin lijin@redhat.com
